src/manifolds/QuotientManifold.jl

```diff
@@ -0,0 +1,165 @@
+
+@doc raw"""
+    canonical_project(M, p)
+
+compute the canonical projection ``π`` on a manifold ``\mathcal M`` that
+[`IsQuotientManifold`](@ref), e.g. a [`QuotientManifold`](@ref).
+The canonical (or natural) projection ``π`` from the total space ``\mathcal N``
+onto ``\mathcal M`` given by
+
+```math
+    π = π_{\mathcal N, \mathcal M} : \mathcal N → \mathcal M, p ↦ π_{\mathcal N, \mathcal M}(p) = [p].
+```
+
+in other words, this function implicitly assumes, that the total space ``\mathcal N`` is given,
+for example explicitly when `M` is a [`QuotientManifold`](@ref) and `p` is a point on `N`.
+"""
+function canonical_project(M::AbstractManifold, p)
+    allocate_result(M, canonical_project, p)
+    return canonical_project!(M, q, p)
+end
+
+@doc raw"""
+    canonical_project!(M, q, p)
+
+compute the canonical projection ``π`` on a manifold ``\mathcal M`` that
+[`IsQuotientManifold`](@ref), e.g. a [`QuotientManifold`](@ref) in place of `q`.
+
+See [`canonical_project`](@ref) for more details.
+"""
+canonical_project!(M::AbstractManifold, q, p)
+
+decorated_manifold(M::QuotientManifold) = M.manifold
+
+@doc raw"""
+    differential_canonical_project(M, p, X)
+
+compute the differential of the canonical projection ``π`` on a manifold ``\mathcal M`` that
+[`IsQuotientManifold`](@ref), e.g. a [`QuotientManifold`](@ref).
+The canonical (or natural) projection ``π`` from the total space ``\mathcal N``
+onto ``\mathcal M``, such that its differential
+
+```math
+ Dπ(p) : T_p\mathcal N → T_{π(p)}\mathcal M
+```
+
+where again the total space might be implicitly assumed, or explicitly when using a
+[`QuotientManifold`](@ref) `M`. So here `p` is a point on `N` and `X` is from ``T_p\mathcal N``.
+"""
+function differential_canonical_project(M::AbstractManifold, p, X)
+    allocate_result(M, differential_canonical_project, p, X)
+    return differential_canonical_project!(M, q, p)
+end
+
+@doc raw"""
+    differential_canonical_project!(M, Y, p, X)
+
+compute the differential of the canonical projection ``π`` on a manifold ``\mathcal M`` that
+[`IsQuotientManifold`](@ref), e.g. a [`QuotientManifold`](@ref). See [`differential_canonical_project`](@ref) for details.
+"""
+differential_canonical_project!(M::AbstractManifold, q, p)
+
+@doc raw"""
+    get_total_space(M::AbstractDecoratorManifold)
+
+Return the total space of a manifold that [`IsQuotientManifold`](@ref), e.g.
+a [`QuotientManifold`](@ref).
+"""
+get_total_space(::AbstractManifold)
+get_total_space(M::QuotientManifold) = M.total_space
+
+@doc raw"""
+    get_orbit_action(M::AbstractDecoratorManifold)
+
+Return the group action that generates the orbit of an equivalence class, that is for the case
+that
+
+```math
+\mathcal M = \mathcal N / \mathcal O,
+```
+
+where ``\mathcal O`` is a Lie group with its group action generating the orbit.
+"""
+get_orbit_action(::AbstractManifold)
+
+@doc raw"""
+    horizontal_lift(N, q, X)
+    horizontal_lift(QuotientManifold{M,N}, p, X)
```

Comment:
Why isn't this just `horizontal_lift(N::AbstractManifold, q, X)`?

Reply:
The general problem here is that q is from N but X is from T_pM (p=pi(q)). So as the first line indicates it is what you ask for, but then we again have a lot of implicit assumptions, i.e. that M and pi (canonical project) are clear from context. If that is not the case M should be enough to know additionally (since then we know domain and range of pi for example). So if that is not the case we need the whole `QuotientManifold{M,N}` instead of just `N`.

But as I said it _is_ just what you say is we have a lot a lot of implicit assumptions.

Reply:
```suggestion
    horizontal_lift(N::AbstractManifold, q, X)
    horizontal_lift(::QuotientManifold{𝔽,MT<:AbstractManifold{𝔽},NT<:AbstractManifold}, p, X) where {𝔽}
```
OK, so this I think would be more clear?

Reply:
Sure, that is more detailed, you‘re right.